### ISSUE TITLE
Rule proposal: Return before defer

### ISSUE THREAD

**Issue by reitzig** [realm/SwiftLint#1358]
I had code of the form:

```swift
let semaphore: DispatchSemaphore(value: 0)
asyncCall(..., completionHandler: { _ in
	if flag { return }
	defer { semaphore.signal() }
	...
})
semaphore.wait()
```

I added the early return some while after the `defer`, and it never triggered (it's a rather obscure check in test code). This, of course, livelocks if flag is true -- a small oversight, but it took a while to find.

The problem is that at first sight, the code is fine: `defer` makes sure that no matter how the function returns, the semaphore is signalled -- well, only those returns *below*.

Hence, I propose a warning for returns that come before defers, or the other way around.

**Comment by marcelofabri** [comment]
I'm not sure this is general enough to be a rule, as there're valid reasons to return before a `defer`. For example, you may not even want to create the variables that will be used inside `defer`.

Anyway, let's see how other people feel about this.

**Comment by reitzig** [comment]
I agree -- it's no per se a problem, only a likely (?) source of mistakes. Definitely opt-in!

**Comment by SDGGiesbrecht** [comment]
### Explanation of `defer`

`defer` is intended so that paired operations can be written together instead of bracketing large amounts of code between them.

It allows the reduction of this mess...

```swift
func doNetworking() throws -> [Message] {

    let socket = try Socket()

    do {
        try socket.schedule()
    } catch let error {
        socket.invalidate()
        throw error
    }
    
    do {
        try socket.startListening()
    } catch let error {
        socket.unschedule()
        socket.invalidate()
        throw error
    }

    do {
        try socket.openConnections()
    } catch let error {
        socket.stopListening()
        socket.unschedule()
        socket.invalidate()
        throw error
    }

    while socket.notFinished {}

    let result = try socket.messagesReceived
    
    socket.closeConnections()
    socket.stopListening()
    socket.unschedule()
    socket.invalidate()

    return result
}
```

...to this far clearer code...

```swift
func doNetworking() throws -> [Message] {

    let socket = try Socket()
    defer { socket.invalidate() }

    try socket.schedule()
    defer { socket.unschedule() }

    try socket.startListening()
    defer { socket.stopListening() }

    try socket.openConnections()
    defer { socket.closeConnections() }

    while socket.notFinished {}

    return try socket.messagesReceived
}
```

Note that if `try socket.schedule()` throws an error, the code never gets to `try socket.openConnections()`, so it may violate a precondition to call `socket.closeConnections()`. That is why `defer` only triggers if the code reaches its location. In the case of an error in `try socket.schedule()`, code has only passed through the first `defer`, so it is the only one executed when the function returns.

***

### `return` before `defer`

Imagine we did some caching in the above example. Then it would look like this:

```swift
private let cache = [Message]?
func doNetworking() throws -> [Message] {

    if let result = cache {
        return result
    }

    let socket = try Socket()
    defer { socket.invalidate() }

    try socket.schedule()
    defer { socket.unschedule() }

    try socket.startListening()
    defer { socket.stopListening() }

    try socket.openConnections()
    defer { socket.closeConnections() }

    while socket.notFinished {}

    let result = try socket.messagesReceived
    cache = result
    return result
}
```

Note that it would be a *very* bad idea to move the `defer` statements up above the first return. If we did...

```swift
private let cache = [Message]?
func doNetworking() throws -> [Message] {

    defer { socket.invalidate() }
    defer { socket.unschedule() }
    defer { socket.stopListening() }
    defer { socket.closeConnections() }

    if let result = cache {
        return result
    }

    let socket = try Socket()
    try socket.schedule()
    try socket.startListening()
    try socket.openConnections()

    while socket.notFinished {}

    let result = try socket.messagesReceived
    cache = result
    return result
}
```

...an error in `try Socket()` would stop code from ever getting to `try socket.schedule()`, but it would still trigger *all* the `defer` statements above it, none of which are safe to do. We would quite likely have a `BAD_ACCESS` avalanche on our hands.
